src/sensors/PointCloud2.js

```diff
@@ -80,7 +80,8 @@ ROS3D.PointCloud2.prototype.subscribe = function(){
   this.rosTopic = new ROSLIB.Topic({
     ros : this.ros,
     name : this.topicName,
-    messageType : 'sensor_msgs/PointCloud2'
+    messageType : 'sensor_msgs/PointCloud2',
+    compression: 'cbor'
```

Comment:
Can we have this configurable as a parameter in `options` in the constructor? This way it can be set by the user. Might be good to default to no compression to maintain backwards compatibility.

Which makes me wonder why this was not using png compression as default in the first place. In an adapted fork I have good experience with png compression with quite big clouds.

Reply:
I don't see any reason to make this configurable.  None of the other ROS3D underlying subscriptions have configurable compression.

If "cbor" compression is requested but not available on the server, it will use JSON, so it's already backwards compatible.

Reply:
When the cbor compression PR lands in `rosbridge` we will have 3 options: none (JSON), png, and cbor. I have no problem with making the default `cbor` if omitted (and if it proves robust and bug-free), but I would like the user to have the option for choosing the other 2 types as well.

Reply:
Is there a good reason to use another type of compression?

Reply:
I think it would be reasonable for all the visualizer classes to have their topic options exposed.  I'd like to be able to override `throttle_rate` for example.  Maybe this should be a separate PR?

Reply:
> Is there a good reason to use another type of compression?

I think you yourself summarized it pretty well in https://github.com/RobotWebTools/rosbridge_suite/issues/367 :)

> ### When to use CBOR
> CBOR is not ideal for all messages and transports, but it's safe to say that it's the best choice for any message containing a large amount of binary data when using roslibjs in a browser.
> 
> JSON is still faster and better for small messages such as transforms, or messages containing mostly strings.
> 
> PNG is best in some specific edge cases. It is still potentially the smallest wire size where transport compression is not available, depending on the data. When the data is mostly redundant, e.g. an array full of zeros, PNG messages can be smaller and decode faster than CBOR, because the deflate algorithm is efficient in this case, but when the data is more random, PNG will often have a larger wire size and longer decoding time than regular JSON compression. This makes it a poor choice for PointCloud2 and situationally useful for Image or OccupancyGrid.

---

> I think it would be reasonable for all the visualizer classes to have their topic options exposed. I'd like to be able to override `throttle_rate` for example. Maybe this should be a separate PR?

Fine by me either way, but especially since exposing compression methods is of the same nature it can be part of this PR.

Reply:
Yes, this is a case where CBOR is always the best option.

Reply:
In my testing on a 5,000 point cloud, the byte size of cbor and png messages is roughly equivalent (120k vs. ~117k).

The png message is a base64 encoding of a png of a base64 encoding of a float32 bytestream.  Each time you base64 you increase the size of the data by 33%.  Floating point numbers don't typically have enough redundancy for deflate compression to get value.  So you end up burning Python and JavaScript time for a marginal reduction in size.

Reply:
Tshark captures help us compare the wire size when using the new WebSocket per-message deflate feature of rosbridge_websocket.

CBOR no deflate (120534):
```
  413 13.099337807          ::1 → ::1          TCP 65550 9090 → 47422 [ACK] Seq=10712414 Ack=1246 Win=49152 Len=65464 TSval=3162249695 TSecr=3162249601 [TCP segment of a reassembled PDU]
  414 13.099364625          ::1 → ::1          WebSocket 54984 WebSocket Binary [FIN] 
```

PNG no deflate (117379):
```
  295 9.364413005          ::1 → ::1          TCP 65550 9090 → 47630 [ACK] Seq=10197181 Ack=1245 Win=49152 Len=65464 TSval=3162397951 TSecr=3162397850 [TCP segment of a reassembled PDU]
  296 9.364464726          ::1 → ::1          WebSocket 51829 WebSocket Text [FIN]
```

CBOR with deflate (60410):
```
  231 11.311518921          ::1 → ::1          WebSocket 60410 WebSocket Binary [FIN]
```

PNG with deflate (89020):
```
  306 9.741264070          ::1 → ::1          TCP 65550 9090 → 47870 [ACK] Seq=8081076 Ack=807 Win=45952 Len=65464 TSval=3162617134 TSecr=3162617034 [TCP segment of a reassembled PDU]
  307 9.741284585          ::1 → ::1          WebSocket 23470 WebSocket Text [FIN]
```

Reply:
So maybe a bytestream of floating point numbers does have enough redundancy for deflate to get value, but not if it is base64 encoded first.

For clients and/or transports that do not support binary messages or per-message deflate, it's possible that we could get good results png encoding a CBOR payload instead of JSON, "cbor_png" compression.  This is a good experiment for another day.

Reply:
It's still possible to have a broken combination:

* ros3djs requesting "cbor" compression
* old version of roslibjs which can't decode "cbor" will still pass on "cbor" to the server
* new version of rosbridge which supports "cbor"

So I'll add the optional compression soon and maybe fix roslibjs to not request compression types from the server that it can not decode.

Reply:
@mvollrath Hi, have you tried visualization_msgs/MarkerArray from octomap_server. It is a bunch of cubes. 
My test result with MarkerArray is different from Pointcloud2.   The performance for ros3djs when receiving MarkerArray is not acceptable.

Reply:
CBOR is bad for MarkerArray iirc, because most of the data is string keys and there's not a lot of content.  JSON is generally faster for messages with a lot of different fields, CBOR is generally faster for messages with a few fields and large numeric arrays.  Using no roslib compression (plain JSON) is probably best, and I recommend setting up rosbridge to use websocket compression.

Reply:
Thanks for the answer. It's very useful to me:D